packages/mobx/src/types/observableobject.ts

```diff
@@ -139,116 +167,423 @@ export class ObservableObjectAdministration
             if (notify) notifyListeners(this, change)
             if (__DEV__ && notifySpy) spyReportEnd()
         }
+        return true
     }
 
-    has_(key: PropertyKey) {
-        const map = this.pendingKeys_ || (this.pendingKeys_ = new Map())
-        let entry = map.get(key)
-        if (entry) return entry.get()
-        else {
-            const exists = !!this.values_.get(key)
-            // Possible optimization: Don't have a separate map for non existing keys,
-            // but store them in the values map instead, using a special symbol to denote "not existing"
+    get_(key: PropertyKey): any {
+        if (globalState.trackingDerivation && !hasProp(this.target_, key)) {
+            // Key doesn't exist yet, subscribe for it in case it's added later
+            this.has_(key)
+        }
+        return this.target_[key]
+    }
+
+    /**
+     * @param {PropertyKey} key
+     * @param {any} value
+     * @param {Annotation|boolean} annotation true - infer from descriptor, false - copy as is
+     * @param {boolean} proxyTrap whether it's called from proxy trap
+     * @returns {boolean|null} true on success, false on failure (proxyTrap + non-configurable), null when cancelled by interceptor
+     */
+    set_(key: PropertyKey, value: any, proxyTrap: boolean = false): boolean | null {
+        // Don't use .has(key) - we care about own
+        if (hasProp(this.target_, key)) {
+            // Existing prop
+            if (this.values_.has(key)) {
+                // Observable (can be intercepted)
+                return this.setObservablePropValue_(key, value)
+            } else if (proxyTrap) {
+                // Non-observable - proxy
+                return Reflect.set(this.target_, key, value)
+            } else {
+                // Non-observable
+                this.target_[key] = value
+                return true
+            }
+        } else {
+            // New prop
+            return this.extend_(
+                key,
+                { value, enumerable: true, writable: true, configurable: true },
+                this.defaultAnnotation_,
+                proxyTrap
+            )
+        }
+    }
+
+    // Returns false for non-enumerable by design!
```

Comment:
The methods on administration now match proxy traps and more or less object api (`remove` ~ `delete`). So callsites are  `object-api.has` and `handler.has` (being a trap for `in` - it's actually mentioned in the comment few lines below).
However the whole thing is actually quite complicated and possibly broken under specific conditions. 
Currently we have a map of `pendingKey`s, which is an `ObservableValue<boolean>` representing whether the key exists or not.
But the *existence of key* is defined differently for different operations:
`o.get(key)` - exists if own or inherited, enumerability doesn't matter
`o.getOwnPropertyDescriptor(key)` - exists if own, enumerability doesn't matter
`key in o` - exists if own or inherited and enumerable
So in theory, if we want to allow up-ahead subsciptions for all of these operations, we would need 3 different `pendingKeys` per acessed key.
Alternatively we could replace the `ObservableValue<boolean>` by simple `Atom` and `reportChanged` anytime any `key` is added/removed, regardles of enumerable/own/inherited. That would cause over-reporting in certain cases, but shouldn't matter most of time.
By *"most of the time"* I mean you either use the object as a simple data structure where you manipulate only own+enumerable, or you have some smarter object which usually has a fixed shape.

Reply:
@urugator just went through the code. I don't see any obvious mistakes in it any more. The reactivity is in principle limited to what happens on *own*, but it results properly reflect the presence of the key in prototype as well, and enumerability is not part of the consideration. LGTM